ferrilata-devilline/Repositories/LevelRepository.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using ferrilata_devilline.Models.DAOs;
+using Microsoft.EntityFrameworkCore;
+
+namespace ferrilata_devilline.Repositories
+{
+    public class LevelRepository : ILevelRepository
+    {
+        private readonly ApplicationContext _applicationContext;
+
+        public LevelRepository(ApplicationContext applicationContext)
+        {
+            _applicationContext = applicationContext;
+        }
+        public void SaveOrUpdate(Level level)
+        {
+            if (_applicationContext.Levels.Find(level.LevelId) == null)
+            {
+                _applicationContext.Levels.Add(level);
+            }
+
+            _applicationContext.Update(level);
+            _applicationContext.SaveChanges();
+        }
+
+        public List<Level> RetrieveLevelsFromDB()
+        {
+            return _applicationContext.Levels.Include(level => level.UserLevels.Select(x => x.User)).ToList();
+        }
+
+        public Level FindLevelById(long id)
+        {
+            return _applicationContext.Levels.SingleOrDefault(x => x.LevelId == id);
+        }
+
+        public void DeleteLevelById(long id)
+        {
+            _applicationContext.Levels.Remove(FindLevelById(id));
```

Comment:
This is not a big problem but a thing to keep in mind. You are calling a method and in the parameter you call another one. Of course it works fine but the best practice is to call the FindLevelById before and save it to a variable. If a bug will happen it's much easier to debug the code when every line is responsible for one action. Saving an additional line in the method just not worth the trouble it may cause.